### ISSUE TITLE
decode_cf not concatenating string arrays

### ISSUE THREAD

**Issue by mogismog** [pydata/xarray#862]
**TL;DR**: `xarray.conventions.decode_cf()` doesn't seem to want to concatenate string arrays after opening up dataset with `decode_cf=False`.

**OS**: Tried on both OS X 11.10 and 11.11
**xarray version**: 0.7.2 installed via conda
**Python version**: 2.7.11

Hey all,

I'm not sure if this is a bug or the intended behavior, but running `xarray.conventions.decode_cf` doesn't seem to concatenate 2D string arrays as promised given certain circumstances.

Specifically,  MADIS netCDF files have `_FillValue`/`missing_value` conflicts. When opening up the file in `xarray`, the exception gives this suggestion:

`ValueError: ('Discovered conflicting _FillValue and missing_value.  Considering opening the offending dataset using decode_cf=False, corrected the attributes', 'and decoding explicitly using xarray.conventions.decode_cf(ds)')`

Doing this, though, doesn't result in 2D string arrays being concatenated:

```
In [50]: import xarray as xr

In [51]: fname = '20160518_1200'

In [52]: ds = xr.open_dataset(fname, decode_cf=False)

In [53]: ds.stationId
Out[53]:
<xarray.DataArray 'stationId' (recNum: 126154, maxStaIdLen: 6)>
[756924 values with dtype=|S1]
Coordinates:
  * maxStaIdLen  (maxStaIdLen) int64 0 1 2 3 4 5
  * recNum       (recNum) int64 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 ...
Attributes:
    long_name: alphanumeric station Id
    reference: station table

In [54]: for _, v in ds.variables.iteritems():
    _fix_fillval_conflict(v) # You can find this function in the linked gist
   ....:

In [55]: decoded_ds = xr.conventions.decode_cf(ds, concat_characters=True)

In [56]: decoded_ds.stationId
Out[56]:
<xarray.DataArray 'stationId' (recNum: 126154, maxStaIdLen: 6)>
[756924 values with dtype=|S1]
Coordinates:
  * maxStaIdLen  (maxStaIdLen) int64 0 1 2 3 4 5
  * recNum       (recNum) int64 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 ...
Attributes:
    long_name: alphanumeric station Id
    reference: station table
```

That said, if you pass `decode_cf=True` and option to not do things like `mask_and_scale` and `decode_times` (due to the aforementioned conflict), the string arrays get concatenated:

```
In [57]: ds = xr.open_dataset(fname, decode_cf=True, mask_and_scale=False, decode_times=False)

In [58]: ds.stationId
Out[58]:
<xarray.DataArray 'stationId' (recNum: 126154)>
[126154 values with dtype=|S6]
Coordinates:
  * recNum   (recNum) int64 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 ...
Attributes:
    long_name: alphanumeric station Id
    reference: station table
```

We then can fix the conflict and run `decode_cf` without issue. This is an acceptable (albeit not immediately obvious or intuitive) workaround, of course, but I'm not sure if this behavior is known or intended. I'll fully admit that it may be an issue with the specific netCDF data I'm working with (NOAA MADIS data, which is kind of a train wreck w/r/t CF convention to begin with), but I don't  have any other datasets with which to test.

To that extent, I've [coded up tests](https://gist.github.com/mogismog/4ff94dd3d67afd612520bc0755f614ca) and [uploaded a gzipped MADIS netCDF file to DropBox](https://www.dropbox.com/s/gvisjl2skpf4zzr/20160518_1200.gz?dl=1) if you're interested in reproducing this behavior.

Thanks!

**Comment by shoyer** [comment]
The heuristic we use for determining if you can concatenate over a dimension includes checking if it's included as a variable: https://github.com/pydata/xarray/blob/v0.7.2/xarray/conventions.py#L800

If a potential dummy dimension is also a variable, we don't concatenate over it.

Of course, every dimension gets a variable by the time you've turned it into a dataset, so this never works on datasets, only data stores.

I'm certainly open to ideas on how to improve this. Possibly accepting an explicit lists of dimensions to concatenate over (and remove) would be the way to go.

**Comment by mogismog** [comment]
Hey @shoyer,

Sorry for the delayed response. Passing a list of dimensions over which to concatenate over seems like it would be the easiest workaround with the fewest questions asked. As you mentioned, every dimension gets a variable by the time it is a dataset, so another option (that I'll admit I haven't thought all the way through and may not even work) would be to first check if `decode_cf` is working on a `Dataset` or `AbstractDataStore`, and then decide whether to concatenate over a dimension or not. I could see the latter idea not working out so well, but I'd be curious about your thoughts.

Either way, I can put something together this week and open up a PR.

**Comment by shoyer** [comment]
>  first check if decode_cf is working on a Dataset or AbstractDataStore (e: which it already does anyway), and then decide whether to concatenate over a dimension or not

This seems a little too magical to me. How would we know if the dataset dimension was added intentionally or not?

**Comment by mogismog** [comment]
> This seems a little too magical to me. How would we know if the dataset dimension was added intentionally or not?

Yeah, that's a fair point. I'll put together something that uses an optional list of dimensions to concatenate over. Thanks!
